### PR TITLE
add regression test for key backup bug

### PR DESCRIPTION
**Pull request by majestrate**
make sure the bug is squashed with a regression test.

### CODE REVIEW

File: llarp/config/config.hpp [hunk 1]
```diff
@@ -72,11 +72,11 @@ namespace llarp
...
 
-    std::string m_keyfile;
+    std::optional<fs::path> m_keyfile;
     std::string m_endpointType;
     bool m_reachable = false;
-    int m_hops = -1;
-    int m_paths = -1;
+    std::optional<int> m_Hops;
+    std::optional<int> m_Paths;
     bool m_AllowExit = false;
```
Comment (notlesh) [inline llarp/config/config.hpp:80]: Did this fix something, or is it more of a clean up thing?

File: llarp/util/logging/logger.hpp [hunk 2]
```diff
@@ -60,9 +60,25 @@ namespace llarp
...
+  {
+    LogSilencer();
+    ~LogSilencer();
+    explicit LogSilencer(LogContext& ctx);
+
+   private:
+    LogContext& parent;
+    ILogStream_ptr stream;
+  };
+
```
Comment (notlesh) [inline llarp/util/logging/logger.hpp:75]: this is a cool use of RAII :)

File: test/regress/2020-06-08-key-backup-bug.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  config->impl.bootstrap.skipBootstrap = true;
+  auto ptr = llarp_main_init_from_config(config, false);
+  llarp_config_free(config);
+  return ptr;
+}
+
+TEST_CASE("key backup bug regression test", "[regress]")
+{
+  llarp::LogSilencer shutup;
+  for (const fs::path& path : {"regress-1.private", "regress-2.private", ""})
```
Comment (notlesh) [inline test/regress/2020-06-08-key-backup-bug.cpp:23]: Some comments about what these nested loops are doing would help

File: llarp/config/config.hpp (hunk 1 above)
Comment (majestrate) [reply llarp/config/config.hpp:80]: it's a cleanup

**Comment by majestrate** [comment]
> It's not clear from the regression test what the bug is or what the test is trying to do. I'm guessing the bug had something to do with an invalid path or empty string being used as the keyfile?

this was the missing/invalid self.signed reseting the hidden service keys bug, i wanted to add a regression test to make sure it's gone.

File: test/regress/2020-06-08-key-backup-bug.cpp (hunk 3 above)
Comment (majestrate) [reply test/regress/2020-06-08-key-backup-bug.cpp]: added comments

**Pull request merged**